### PR TITLE
planner: use TableInfo.DBID to locate schema

### LINKED ISSUE #57779: It takes too long time for query tables with tidb_table_id
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

1.5M databases, 3M tables 
We should optimize the execution time of the follow query
```sql
mysql> select TABLE_SCHEMA from INFORMATION_SCHEMA.tables where tidb_table_id=200461;
+--------------+
| TABLE_SCHEMA |
+--------------+
| sb10wnormal1 |
+--------------+
1 row in set (14 min 21.65 sec)
```

### 4. What is your TiDB version? (Required)

master
Release Version: v8.5.0-alpha-230-gca395fa4be
Edition: Community
Git Commit Hash: ca395fa4be5278acbdec0b15803e26877bee66c0
Git Branch: HEAD
UTC Build Time: 2024-11-27 13:21:51
GoVersion: go1.23.3
Race Enabled: false
Check Table Before Drop: false
Store: tikv

### LINKED ISSUE #57783: It takes too long time for add index in million tables
## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

3M tables
Add index for a table with 10k rows takes about 20 minutes
```sql
   JOB_ID: 7504857
     DB_NAME: sbtest29
  TABLE_NAME: sbtest2
    JOB_TYPE: add index
SCHEMA_STATE: public
   SCHEMA_ID: 3324475
    TABLE_ID: 1000158
   ROW_COUNT: 20000
 CREATE_TIME: 2024-11-28 02:28:25.788000
  START_TIME: 2024-11-28 03:13:49.938000
    END_TIME: 2024-11-28 03:33:51.738000
       STATE: synced
    COMMENTS: ingest, DXF
```
### 4. What is your TiDB version? (Required)

Release Version: v8.5.0
Edition: Community
Git Commit Hash: 3da59f7067f3188f33117164572d2fe449c63249
Git Branch: HEAD
UTC Build Time: 2024-11-22 14:46:47
GoVersion: go1.23.3
Race Enabled: false
Check Table Before Drop: false
Store: tikv

### PR DESCRIPTION
**Pull request by tangenta**
### What problem does this PR solve?

Issue Number: close #57783, close #57779

Problem Summary:

When `tableInfo` is located, it is unnecessary to read all databases to match.

### What changed and how does it work?

This PR changes to use `TableInfo.DBID` to locate the schema.

### Check List

Tests 

- [ ] Unit test
- [x] Integration test
- [x] Manual test (add detailed scripts or steps below)
  Use the same environment mentioned in #57783:
  ```sql
  mysql> set @@tidb_mem_quota_query = 10*1024*1024*1024;
  Query OK, 0 rows affected (0.22 sec)
  
  mysql> select TABLE_SCHEMA, table_name from INFORMATION_SCHEMA.tables where tidb_table_id=1000470;
  +--------------+------------+
  | TABLE_SCHEMA | table_name |
  +--------------+------------+
  | sbtest99     | sbtest2    |
  +--------------+------------+
  1 row in set (0.05 sec)
  
  mysql> select count(1) from sbtest99.sbtest2;
  +----------+
  | count(1) |
  +----------+
  |    20000 |
  +----------+
  1 row in set (0.08 sec)
  
  mysql> alter table sbtest99.sbtest2 add index idx(ac);
  Query OK, 0 rows affected (3.06 sec)
  ```
  After this PR, it takes 
  - 0.05s to query table that filtered with tidb_table_id.
  - 3.06s to add index for a table with 20k rows.
- [ ] No need to test
  > - [ ] I checked and no code files have been changed.
  > 

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 1]
```diff
@@ -720,26 +733,27 @@ func listTablesForEachSchema(
...
-	ctx context.Context,
 	is infoschema.InfoSchema,
-	schemas []pmodel.CIStr,
+	extractedSchemas []pmodel.CIStr,
+	unspecified bool,
 	tableSlice []*model.TableInfo,
 ) ([]pmodel.CIStr, []*model.TableInfo, error) {
 	schemaSlice := make([]pmodel.CIStr, 0, len(tableSlice))
 	for i, tbl := range tableSlice {
+		dbInfo, ok := is.SchemaByID(tbl.DBID)
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:743]: Seem like `DBID` field is not persisted, can we use this field to do search?

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 2]
```diff
@@ -112,22 +112,32 @@ func (e *InfoSchemaBaseExtractor) GetBase() *InfoSchemaBaseExtractor {
...
 		})
-		return filterSchemaObjectByRegexp(e, ec.schema, ret, extractStrCIStr)
+		return filterSchemaObjectByRegexp(e, e.extractableColumns.schema, ret, extractStrCIStr)
+	}
+	return extractedSchemas
+}
+
+func (e *InfoSchemaBaseExtractor) listExtractedSchemas(
+	is infoschema.InfoSchema,
+) (schemas []pmodel.CIStr, unspecified bool) {
```
Comment (tangenta) [inline pkg/planner/core/memtable_infoschema_extractor.go:128]: Although we can use `schemas == nil` to indicate no predicate is specified, developers may get confused to distinguish from `len(schemas) == 0`.

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 3]
```diff
@@ -720,26 +733,27 @@ func listTablesForEachSchema(
...
+	unspecified bool,
 	tableSlice []*model.TableInfo,
 ) ([]pmodel.CIStr, []*model.TableInfo, error) {
 	schemaSlice := make([]pmodel.CIStr, 0, len(tableSlice))
 	for i, tbl := range tableSlice {
+		dbInfoName, ok := is.SchemaNameByTableID(tbl.DBID)
+		if !ok {
+			continue
+		}
+		if len(extractedSchemas) == 0 && unspecified {
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:747]: Seems like if `unspecified == true`, then `len(extractedSchemas)` is always `nil`.

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 4]
```diff
@@ -150,7 +159,8 @@ func (e *InfoSchemaBaseExtractor) ListSchemasAndTables(
 			findTablesByID(is, tableIDs, tableNames, tableMap)
 			tableSlice := maps.Values(tableMap)
 			tableSlice = filterSchemaObjectByRegexp(e, ec.table, tableSlice, extractStrTableInfo)
-			return findSchemasForTables(ctx, is, schemas, tableSlice)
+			schemas, unspecified := e.listExtractedSchemas(is)
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:162]: By the way, it's a bit weird to check predicates by `ec.tableID != ""`

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 5]
```diff
@@ -112,22 +112,32 @@ func (e *InfoSchemaBaseExtractor) GetBase() *InfoSchemaBaseExtractor {
...
 		slices.SortFunc(ret, func(a, b pmodel.CIStr) int {
 			return strings.Compare(a.L, b.L)
 		})
-		return filterSchemaObjectByRegexp(e, ec.schema, ret, extractStrCIStr)
+		return filterSchemaObjectByRegexp(e, e.extractableColumns.schema, ret, extractStrCIStr)
+	}
+	return extractedSchemas
+}
+
+func (e *InfoSchemaBaseExtractor) listExtractedSchemas(
```
Comment (joechenrh) [inline pkg/planner/core/memtable_infoschema_extractor.go:126]: Can we add comment for this function

File: pkg/planner/core/memtable_infoschema_extractor.go (hunk 4 above)
Comment (tangenta) [reply pkg/planner/core/memtable_infoschema_extractor.go:162]: Do you mean we should decide this at the compile time? Do you have any suggestions?

Comment (joechenrh) [reply pkg/planner/core/memtable_infoschema_extractor.go]: Nevermind, because I have no suggestion yet :(

**Review by joechenrh — APPROVED** [review]

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 6]
```diff
@@ -723,23 +737,25 @@ func listTablesForEachSchema(
...
 	is infoschema.InfoSchema,
-	schemas []pmodel.CIStr,
 	tableSlice []*model.TableInfo,
 ) ([]pmodel.CIStr, []*model.TableInfo, error) {
+	schemas, unspecified := e.listPredicateSchemas(is)
 	schemaSlice := make([]pmodel.CIStr, 0, len(tableSlice))
 	for i, tbl := range tableSlice {
+		dbInfo, ok := is.SchemaByID(tbl.DBID)
+		if !ok {
+			continue
```
Comment (tiancaiamao) [inline pkg/planner/core/memtable_infoschema_extractor.go:749]: Something is wrong? how about add some log in this situation?

**Review by tiancaiamao — APPROVED** [review]

File: pkg/planner/core/memtable_infoschema_extractor.go [hunk 7]
```diff
@@ -160,13 +175,15 @@ func (e *InfoSchemaBaseExtractor) ListSchemasAndTables(
...
+			return findSchemasForTables(e, is, tableSlice)
 		}
 	}
 	if len(tableNames) > 0 {
 		tableNames = filterSchemaObjectByRegexp(e, ec.table, tableNames, extractStrCIStr)
+		schemas := e.ListSchemas(is)
 		return findTableAndSchemaByName(ctx, is, schemas, tableNames)
 	}
+	schemas := e.ListSchemas(is)
 	return listTablesForEachSchema(ctx, e, is, schemas)
```
Comment (Rustin170506) [inline pkg/planner/core/memtable_infoschema_extractor.go:187]:
```suggestion
	schemas := e.ListSchemas(is)
	if len(tableNames) > 0 {
		tableNames = filterSchemaObjectByRegexp(e, ec.table, tableNames, extractStrCIStr)
		return findTableAndSchemaByName(ctx, is, schemas, tableNames)
	}
	return listTablesForEachSchema(ctx, e, is, schemas)
```

**Review by Rustin170506 — APPROVED** [review]
Looks good to me. Thanks! :shipit:

**Pull request merged**